README.md
# Katanya Internet Positif
[![license](https://img.shields.io/github/license/ditatompel/Katanya-Internet-Positif.svg)](LICENSE)
[![Website](https://img.shields.io/website-up-down-green-red/http/trustpositif.kominfo.go.id.svg)](http://trustpositif.kominfo.go.id)

`Katanya Internet Positif` adalah script sederhana untuk mengambil data blacklist dari trustpositif.kominfo.go.id, melakukan filterisasi kembali untuk duplikasi domain atau alamat wildcard IP yang tidak valid, dan melakukan convert untuk konfigurasi BIND DNS zone blacklist. Untuk informasi lebih lanjut, baca [halaman Wiki](https://github.com/ditatompel/Katanya-Internet-Positif/wiki).

## Petunjuk Penggunaan
Script ini dapat dijalankan menggunakan Python 2.7 atau lebih.

Untuk yang masih menggunakan Python 2.6, masih dapat menjalankan script ini
dengan menginstall `python-argparse` sebagai module tambahan.

### 1st run / update RAW_DATA
Untuk yang **pertama kali menjalankan script ini** (atau ingin melakukan update RAW_DATA), gunakan opsi `-f` atau `--fetch` untuk
mengambil RAW_DATA dari trustpositif.kominfo.go.id

    $ python ./internetpositif.py --fetch

### Bersihkan duplikasi Data / list alamat IP yang tidak valid
Untuk membersihkan duplikasi data, menghapus alamat IP yang tidak valid, gunakan opsi `-c` atau `--clean`

    $ python ./internetpositif.py -c

### BIND zone file generator
Untuk membuat konfigurasi BIND blacklist zone, gunakan opsi `-g` atau `--generate` dengan opsi tambahan `-n` / `--nameserver`, `-e` / `--email`, `-d` / `--domain`

Misalnya :

    $ python ./internetpositif.py --generate -n ns.domain.com -e admin.domain.com -d blokir.domain.com
akan menghasilkan file :

    $TTL 86400      ; 1 day
    @   IN      SOA ns.domain.com. admin.domain.com. (
                                    2017040503 ; serial
                                    3600       ; refresh (1 hour)
                                    120        ; retry (2 minute)
                                    604800     ; expire (1 week)
                                    86400      ; minimum (1 day)
                                    )
            IN      NS      ns.domain.com.
    domain.yang.diblokir.com IN CNAME blokir.domain.com.
    domain.lainnya.yang.diblokir.net IN CNAME blokir.domain.com.
    [dan seterusnya]

### Jalankan semua dalam 1 baris
    $ python ./internetpositif.py -cfg -n=ns.domain.com --email=admin.domain.com -d blokir.domain.com

### Default value
Opsi tertentu memiliki default value, seperti :
* `-n` / `--nameserver` = `localhost.local`
* `-e` / `--email` = `administrator.localhost.local`
* `-d` / `--domain` = `internetbaik.wds.co.id`

Informasi lebih lanjut

    $ python ./internetpositif.py

    -h, --help            show this help message and exit
    -f, --fetch           Ambil data blacklist dari trustpositif.kominfo.go.id
    -c, --clean           Bersihkan duplikasi data dari
                        trustpositif.kominfo.go.id
    -g, --generate        Buat file zone file untuk BIND (BIND zone file
                        generator)

    Opsi tambahan untuk BIND zone file generator:
    -n NAMESERVER, --nameserver NAMESERVER
                        Nameserver option, default : localhost.local
    -e EMAIL, --email EMAIL
                        Email option, default : administrator.localhost.local
    -d DOMAIN, --domain DOMAIN
                        CNAME domain tujuan. Default internetbaik.wds.co.id


## Struktur File
Script akan membuat folder tertentu sesuai dengan opsi perintah yang diberikan.

| Perintah | Alias | Folder | Keterangan |
|---|---|---|---|
| -f | --fetch | ./RAW_DATA/ | data asli yang didapat dari trustpositif.kominfo.go.id |
| -c | --clean | ./DATA/ | hasil list domain yang telah dibersihkan dari folder ./RAW_DATA/ |
| -g | --generate | ./conf/ | hasil convert named zone yang didapat dari folder ./DATA/ |

## Berkontribusi
Lihat [CONTRIBUTING.md](CONTRIBUTING.md) untuk informasi lebih detail mengenai
cara berkontribusi pada repositori ini.

## Lisensi
[MIT License](LICENSE)


CHANGELOG.md
# Changelog

## v1.0.5
* Menggabungkan semua list dari kajian, pengaduan dan porn ke ./conf/semua_blacklist_named.conf saat menulis konfigurasi BIND
* Jika proses pembersihan domain kembar terganggu / dibatalkan, file yang sudah
ada sebelumnya tidak hilang / diupdate
* Membersihkan domain tanpa TLD seperti shesafreak. atau ladyboynancy.


internetpositif.py
#!/usr/bin/python
import sys
try:
    import argparse
except ImportError:
    sys.exit('Silahkan install `python-argparse` atau gunakan python >= 2.7 untuk menjalankan script ini')
import os, re, urllib2, json
ver = '1.0.5'
prog_desc = "Simple Kominfo Trust / Internet Positif db fetcher written in Python"
base_url = 'http://trustpositif.kominfo.go.id/files/downloads/index.php'
path_lists = 'path.json'
dl_dir = 'RAW_DATA'
clean_data = 'DATA'
cnf_dir = 'conf'

def halo() :
    print "\n" + "-+-"*26 + "\n\tPython Kominfo Trust / Intetnet Positif Fetcher " + str(ver)
    print "\t" + prog_desc
    print "\tby ditatompel <svcadm@ditatompel.com> <christian.dita@wds.co.id>"
    print "\thttps://devilzc0de.org/forum/user-11541.html"
    print "-+-"*26+"\n"

def clean_domain(government):
    validate_ip = re.compile("^(\*\.)\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$")
    for censorship in government.keys():
        dir_location = dl_dir + '/' + censorship + '/'
        dest_location = clean_data + '/' + censorship + '/'
        if os.path.isdir(dir_location):
            for filename in os.listdir(dir_location):
                lines_seen = set()
                to_write = []
                if not os.path.exists(os.path.dirname(dest_location)):
                    os.makedirs(os.path.dirname(dest_location))
                for line in open(dir_location + filename, "rb"):
                    # added .replace(' ', "\n") because kominfo source data
                    # contain 2 domains in 1 line -_-"
                    line = line.strip().replace("\r", '').replace(' ', "\n")
                    # discard invalid domain like shesafreak. , ladyboynancy.
                    # or invalid wildcard ip addr like *.200.10.104.174
                    if line.endswith('.') or validate_ip.match(line):
                        print 'Domain / alamat IP tidak valid ditemukan : ' + line
                    else:
                        if line not in lines_seen:
                            lines_seen.add(line)
                            to_write.append(line)
                        else:
                            print 'Data kembar ditemukan : ' + line
                if len(to_write):
                    outfile = open(dest_location + filename, "wb")
                    for domain in to_write:
                        outfile.write(domain + "\n")
                    outfile.close()
                    print str(len(to_write)) + ' domain ditulis di ' + dest_location + filename
        else:
            print 'Direktori ' + dir_location + ' tidak ditemukan'

def fetch_data(government):
    for censorship in government.keys():
        fetch_link = urllib2.urlopen(base_url + "?dir=" + government[censorship])
        raw_html = fetch_link.read()
        links = re.findall('href=".*dlf/unknown', raw_html)
        for link in links:
            link = re.findall('\?dir.*" class=', link)
            link = link[0].replace('" class=', '')
            print "Mengunduh " + base_url + link
            stream = urllib2.urlopen(base_url + link)
            domain_data = stream.read()
            files = link.split('=')
            save_to = dl_dir + '/' + censorship + '/' + files[2] + '.txt'
            if not os.path.exists(os.path.dirname(save_to)):
                os.makedirs(os.path.dirname(save_to))
            with open(save_to, "wb") as db:
                db.write(domain_data)

def generate_named(government, ns, email, domain):
    if not os.path.exists(os.path.dirname(cnf_dir + '/')):
        os.makedirs(os.path.dirname(cnf_dir + '/'))
    head = """
$TTL 86400      ; 1 day
@   IN      SOA %s. %s. (
                                2017040503 ; serial
                                3600       ; refresh (1 hour)
                                120        ; retry (2 minute)
                                604800     ; expire (1 week)
                                86400      ; minimum (1 day)
                                )
        IN      NS      %s.
""" % (ns, email, ns)
    print 'Memproses pembuatan config BIND...'
    # found *.{ipaddr} on blacklist. definitely not valid
    ipaddr_regex = re.compile("^(\*\.)?\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$")
    lines_seen = set()
    to_write = []
    for censorship in government.keys():
        dir_location = clean_data + '/' + censorship + '/'
        if os.path.isdir(dir_location):
            for filename in os.listdir(dir_location):
                if filename == 'domains.txt':
                    print 'Menulis konfigurasi BIND ke ' + cnf_dir + '/' + censorship + '_named.conf'
                    outfile = open(cnf_dir + '/' + censorship + '_named.conf', "wb")
                    outfile.write(head)
                    for line in open(dir_location + filename, "rb"):
                        record = line.replace("\r",'').replace("\n", '')
                        if not ipaddr_regex.match(record):
                            outfile.write(record + " IN CNAME " + domain + ".\n")
                            if record not in lines_seen:
                                lines_seen.add(line)
                                to_write.append(line)
                    outfile.close()
        else:
            print 'Direktori ' + dir_location + ' tidak ditemukan'
    if len(to_write):
        outfile = open(cnf_dir + '/semua_blacklist_named.conf', "wb")
        outfile.write(head)
        for data in to_write:
            outfile.write(data.replace("\r",'').replace("\n", '') + " IN CNAME " + domain + ".\n")
        outfile.close()
        print str(len(to_write)) + ' domain ditulis di ' + cnf_dir + '/semua_blacklist_named.conf'

if __name__ == "__main__":
    if not os.path.isfile(path_lists):
        sys.exit('file ' + path_lists + ' tidak ditemukan. Kunjungi https://github.com/ditatompel/Katanya-Internet-Positif')
    nameserver = 'localhost.local'
    email = 'administrator.localhost.local'
    dest_domain = 'internetbaik.wds.co.id'

    opts = argparse.ArgumentParser(
        description=prog_desc,
        epilog="Contoh: python ./" + sys.argv[0] + " -cfg -n ns.domain.com -e admin.domain.com -d blokir.domain.com")
    opts.add_argument("-f", "--fetch", dest="fetch", action="store_true", help="Ambil data blacklist dari trustpositif.kominfo.go.id")
    opts.add_argument("-c", "--clean", dest="clean", action="store_true", help="Bersihkan duplikasi data dari trustpositif.kominfo.go.id")
    opts.add_argument("-g", "--generate", dest="generate", action="store_true", help="Buat file zone file untuk BIND (BIND zone file generator)")
    group = opts.add_argument_group('Opsi tambahan untuk BIND zone file generator')
    group.add_argument("-n", "--nameserver", dest="nameserver", help="Nameserver option, default : " + nameserver)
    group.add_argument("-e", "--email", dest="email", help="Email option, default : " + email)
    group.add_argument("-d", "--domain", dest="domain", help="CNAME domain tujuan. Default " + dest_domain)
    options = opts.parse_args()

    show_help = True
    with open(path_lists) as source:
        government = json.load(source)
    if options.fetch:
        fetch_data(government)
        show_help = False
    if options.clean:
        clean_domain(government)
        show_help = False
    if options.generate:
        if options.nameserver:
            nameserver = options.nameserver
        if options.email:
            email = options.email
        if options.domain:
            dest_domain = options.domain
        generate_named(government, nameserver, email, dest_domain)
        show_help = False
    if show_help:
        halo()
        opts.print_help()
